source/server/http_dynamic_delay_filter.cc

```diff
@@ -0,0 +1,87 @@
+#include "server/http_dynamic_delay_filter.h"
+
+#include <string>
+
+#include "envoy/server/filter_config.h"
+
+#include "external/envoy/source/common/stats/symbol_table_impl.h" // For StatName
+
+#include "server/common.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Nighthawk {
+namespace Server {
+
+HttpDynamicDelayDecoderFilterConfig::HttpDynamicDelayDecoderFilterConfig(
+    nighthawk::server::ResponseOptions proto_config, Envoy::Stats::Scope& stats_scope)
+    : server_config_(std::move(proto_config)), stats_scope_(stats_scope) {}
+
+HttpDynamicDelayDecoderFilter::HttpDynamicDelayDecoderFilter(
+    HttpDynamicDelayDecoderFilterConfigSharedPtr config)
+    : config_(std::move(config)) {}
+
+void HttpDynamicDelayDecoderFilter::onDestroy() {}
+
+Envoy::Http::FilterHeadersStatus
+HttpDynamicDelayDecoderFilter::decodeHeaders(Envoy::Http::RequestHeaderMap& headers, bool) {
+  base_config_ = config_->server_config();
+  const auto* request_config_header = headers.get(TestServer::HeaderNames::get().TestServerConfig);
+  if (request_config_header) {
+    Utility::mergeJsonConfig(request_config_header->value().getStringView(), base_config_,
+                             error_message_);
+  }
+  if (error_message_.has_value()) {
+    decoder_callbacks_->sendLocalReply(
+        static_cast<Envoy::Http::Code>(500),
+        fmt::format("dynamic-delay didn't understand the request: {}", *error_message_), nullptr,
+        absl::nullopt, "");
+    return Envoy::Http::FilterHeadersStatus::StopIteration;
+  }
+  absl::optional<int64_t> delay;
+  if (base_config_.has_static_delay()) {
+    delay = Envoy::Protobuf::util::TimeUtil::DurationToMilliseconds(base_config_.static_delay());
+  } else if (base_config_.has_gauge_based_delay()) {
+    auto& gauge_based_delay = base_config_.gauge_based_delay();
+    const std::string gauge_name = gauge_based_delay.gauge_name();
+    const uint64_t gauge_value =
+        config_->statsScope()
+            .gaugeFromString(gauge_name, Envoy::Stats::Gauge::ImportMode::Uninitialized)
```

Comment:
can you capture the Gauge& in the config so you don't have to do the name-based lookup on every request?

Also, I'd use Scope::findGauge() instead so you can report an error during config if the gauge doesn't exist.

Another thing that's concerning here is due to dynamic vs symbolic stat names. We don't really have a robust mechanism to look up a stat by string in prod now, due to dynamic vs symbolic stat name segments. To do the lookup we need to either know what segments of the stat name are dynamic vs symbolic, or we need to iterate through all stat names, serialize them to strings, and do string compares.

This is likely to work the way you have it because most (but not all) stat names are purely symbolic. But there is no guarantee.

In tests we keep a string->stat map (very expensive) to make this easy. In prod I think we will have to add:

Scope::slowlyFind(Counter|Gauge|Histogram|TextReadout)ByName(absl::string_view name) which takes time O(num stats). We would definitely want to only do that during config time.

What's worse, some of the stats may exist at the time filter is existed; not sure about that. Might have to lazy-init with an atomic protector or something, maybe using AtomicPtr : https://github.com/envoyproxy/envoy/blob/7ea1f24bd53522408d9bc55316dff9ed1701bc73/source/common/common/thread.h#L156